src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Grammar/GrammarGenerator.cs

```diff
@@ -0,0 +1,357 @@
+        private readonly Dictionary<string, List<Production>> _nameToProductions;
+
+        public GrammarGenerator(Tree tree)
+        {
+            // Syntax refers to a special pseudo-element 'Modifier'.  Just synthesize that since
+            // it's useful in the g4 grammar.
+            _nodes = tree.Types.Where(c => c is AbstractNode || c is Node).ToImmutableArray().Add(
+                new Node
+                {
+                    Name = "Modifier",
+                    Children =
+                    {
+                        new Field
+                        {
+                            Type = SyntaxToken,
+                            Kinds = ModifierKeywords.Select(k => new Kind { Name = k }).ToList()
+                        }
+                    }
+                });
+
+            _nameToProductions = _nodes.ToDictionary(n => n.Name, _ => new List<Production>());
+        }
+
+        public string Run()
+        {
+            // Synthesize this so we have a special node that can act as the parent production for
+            // all structured trivia rules.
+            _nameToProductions.Add("StructuredTriviaSyntax", new List<Production>());
+
+            foreach (var node in _nodes)
+            {
+                // If this node has a base-type, then have the base-type point to this node as a
+                // valid production for itself.
+                if (node.Base is string nodeBase && _nameToProductions.TryGetValue(nodeBase, out var baseProductions))
+                    baseProductions.Add(RuleReference(node.Name));
+
+                if (node is Node)
+                {
+                    if (node.Children.Count == 0)
+                        throw new InvalidOperationException(node.Name + " had no children");
+
+                    // Convert a rule of `a: (x | y | z)` into:
+                    // a: x
+                    //  | y
+                    //  | z;
+                    if (node.Children.Count == 1 && node.Children[0] is Field field && field.IsToken)
+                    {
+                        ProcessProductions(node, field.Kinds.Select(k =>
+                            new List<TreeTypeChild> { new Field { Type = SyntaxToken, Kinds = new List<Kind> { k } } }).ToArray());
+                    }
+                    else
+                    {
+                        ProcessProductions(node, node.Children);
+                    }
+                }
+            }
+
+            // The grammar will bottom out with certain lexical productions.  Just emit a few empty
+            // productions in the grammar file indicating what's going on, and making it so that the
+            // g4 file is considered legal (i.e. no rule references names of rules that don't
+            // exist).
+
+            var lexicalProductions = new List<Production> { new Production("/* see lexical specification */") };
+            _nameToProductions.Add("Token", lexicalProductions);
```

Comment:
nit: should this be a constant?

Reply:
i moved away from as many constants as i didn't think they helped with readability.  I still have them for strings that show up a lot.  but not for ones that are in just one place.